crypto/rand/prov_seed.c

```diff
@@ -44,31 +45,13 @@ size_t ossl_rand_get_user_entropy(OSSL_LIB_CTX *ctx,
                                   unsigned char **pout, int entropy,
                                   size_t min_len, size_t max_len)
 {
-    unsigned char *buf;
     EVP_RAND_CTX *rng = ossl_rand_get0_seed_noncreating(ctx);
-    size_t ret;
 
-    if (rng == NULL)
+    if (rng != NULL && evp_rand_can_seed(rng))
```

Comment:
We really should do something similar in the `ossl_rand_get_user_nonce()` as well.

Reply:
This is still sensible I think and it's a straightforward change.

Reply:
> We really should do something similar in the `ossl_rand_get_user_nonce()` as well.

I would have done it right from the beginning (you know I love consistency) and it would have been a straightforward change, if there hadn't been an awkward fracture in the API's consistency: 


This is the signature of the upcall:


```C
size_t ossl_rand_get_user_nonce(OSSL_LIB_CTX *ctx,
                                unsigned char **pout,
                                size_t min_len, size_t max_len,
                                const void *salt, size_t salt_len)
```

and the obvious implementation would have been

```C
size_t ossl_rand_get_user_nonce(OSSL_LIB_CTX *ctx,
                                unsigned char **pout,
                                size_t min_len, size_t max_len,
                                const void *salt, size_t salt_len)
{
   EVP_RAND_CTX *rng = ossl_rand_get0_seed_noncreating(ctx);

    if (rng != NULL && evp_rand_has_nonce(rng))
        return evp_rand_get_nonce(rng, pout, min_len, max_len, salt, salt_len);
    else
        return ossl_rand_get_nonce(ctx, pout, min_len, max_len, salt, salt_len);
}
```
analogous to the user seed implementation.

However, instead of a pair `rand_get_nonce`/`rand_clear_nonce` where the callee allocates and cleans up the memory,


```C
OSSL_CORE_MAKE_FUNC(size_t,rand_get_nonce,
                    (void *vctx, unsigned char **buffer,
                     size_t min_len, size_t max_len,
                     const void *salt, size_t salt_len))
OSSL_CORE_MAKE_FUNC(void,rand_clear_nonce,
                    (void *vctx, unsigned char *buffer, size_t b_len))
```

there is a only a method

https://github.com/openssl/openssl/blob/fac61ea4618c83826b51aebf03cbc2bc3ac7b8c8/include/openssl/core_dispatch.h#L492-L494

which is expected to be called twice, the first time with a NULL buffer to determine the the length of the buffer to be allocated by the caller (note the `unsigned char *out` instead of `unsigned char **pout`). Also, the `salt` and `salt_len` is missing from the arguments.

And that fracture is not easy to fix, because the method is public as `EVP_RAND_nonce()`. :-/

Reply:
see also https://github.com/openssl/openssl/pull/22394#pullrequestreview-1678916810.

Reply:
`ossl_rand_get_user_nonce()` hasn't been in a release and is consequently mutable at least until later today.  Although you're not proposing a change so it might be okay even so.

More over, `ossl_rand_get_user_nonce()` is only ever called by the FIPS provider (and similar 3rd party providers) which will never call `EVP_RAND_nonce()`, so again it should be okay.

Finally, we don't call `EVP_RAND_nonce()` anywhere and implementing it (very) inefficiently might be acceptable.

Reply:
I did not fully understand you answer. Maybe we can talk about this briefly at the end of our OTC meeting?

Reply:
Or before??

Reply:
Essentially, `EVP_RAND_nonce()` is independent of `ossl_rand_get_user_nonce()`.  The API difference doesn't matter IMO.

Reply:
> Or before??

_(That was not possible due to another meeting)_

Reply:
@paulidale and I (and others) had a talk and I'll come with a proposal within the next days.